tools/zarr-cache-stats.py

```diff
@@ -0,0 +1,299 @@
+#!/usr/bin/env python3
+# /// script
+# requires-python = ">=3.11"
+# dependencies = []
+# ///
+
+"""
+This script takes as arguments one or more `dandidav` Heroku logs (assumed to
+be consecutive) and outputs various statistics related to the Zarr
+manifest cache.
+"""
+
+from __future__ import annotations
+from collections import Counter
+from collections.abc import Iterable, Iterator
+from dataclasses import dataclass, field
+from datetime import datetime
+from enum import Enum
+import json
+import logging
+from pathlib import Path
+import re
+import statistics
+import sys
+import textwrap
+
+log = logging.getLogger(__name__)
+
+
+@dataclass
+class HerokuLog:
+    timestamp: datetime
+    target: str
+    message: str
+    line: str
+
+
+@dataclass
+class Events:
+    misses: list[Miss] = field(default_factory=list)
+    killed_misses: list[KilledMiss] = field(default_factory=list)
+    hits: list[Hit] = field(default_factory=list)
+    evictions: list[Eviction] = field(default_factory=list)
+
+
+@dataclass(frozen=True)
+class MissKey:
+    span_id: str
+    manifest_path: str
+
+
+@dataclass
+class MissMoment:
+    timestamp: datetime
+    approx_cache_len: int
+
+
+@dataclass
+class Miss:
+    manifest_path: str
+    start: MissMoment
+    end: MissMoment
+    #: Number of other misses in progress as of the end of this miss
+    parallel: int
```

Comment:
interesting why of interest "as of the end" and not "as of the beginning"?

Reply:
It all comes out the same either way, but I had to pick one.